File: array sol/3 sum closest/three sum closest.cpp
#include<iostream>
#include<vector>
#include<algorithm>
using namespace std;

class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        int n = nums.size();
        int closest = nums[0]+nums[1]+nums[2];

        for(int i=0; i<n-2; i++) {
            int left = i+1, right = n-1;

            while(left<right) {
                int current = nums[i] + nums[left] +nums[right];

                if(abs(current-target) < abs(closest-target)) {
                    closest = current;
               }

               if(current<target) {
                left++;
               }
               else if(current>target) {
                right--;
               }
               else{
                return current;
               }
            }
        }
        return closest;
    }
};

File: array_sol/3 sum closest/three sum closest.py
class Solution:
    def threeSumClosest(self, nums, target):
        n = len(nums)
        if n < 3:
            raise ValueError("Need at least 3 numbers")

        nums.sort()
        closest = nums[0] + nums[1] + nums[2]

        for i in range(n - 2):
            left, right = i + 1, n - 1

            while left < right:
                current = nums[i] + nums[left] + nums[right]

                # Update closest if this sum is nearer to target
                if abs(current - target) < abs(closest - target):
                    closest = current

                # Move pointers based on comparison with target
                if current < target:
                    left += 1
                elif current > target:
                    right -= 1
                else:
                    # exact match, can't do better
                    return current

        return closest
